src/main/ngapp/accio/src/app/mask.directive.ts

```diff
@@ -0,0 +1,142 @@
+import { Directive, HostListener, ElementRef, Input } from '@angular/core';
+import { MagicWandService } from './magic-wand.service';
+
+@Directive({
+  selector: '[appMask]',
+})
+export class MaskDirective {
+  @Input() scale: number;
+  @Input() originalData: ImageData;
+  @Input() scaledData: ImageData;
+  @Input() maskData: ImageData;
+
+  // create direct reference of canvas on editor.html
+  constructor(
+    private canvas: ElementRef<HTMLCanvasElement>, 
+    private magicWandService: MagicWandService,
+  ) { }
+  private ctx: CanvasRenderingContext2D;
+  private maskImage: HTMLImageElement;
+
+  @HostListener('click', ['$event'])
+  onClick(e: MouseEvent) {
+    const xCoord = e.offsetX;
+    const yCoord = e.offsetY;
+    this.ctx = this.canvas.nativeElement.getContext('2d');
+
+    // tolerance will come from user input
+    const tolerance = 20;
+
+    //  TODO-MASK-maskPixels needs to contain cumulative set of all pixels user clicked and added to the mask
+    // returns an array indices of each pixel in the mask
+    const maskPixels = this.magicWandService.floodfill(this.originalData, Math.floor(xCoord / this.scale), Math.floor(yCoord / this.scale), tolerance);
+
+    // alpha value hardcoded for now
+    this.drawMask(maskPixels, this.scaledData, 1);
+  }
+
+  /**
+   *  Creates a UI mask to the user by reading in all values in the given maskPixels
+   *  Clears the canvas fisrt so the mask can be saved as a new image and then over-
+   *    lays the mask on top of the scaled image. 
+   *  Adds all pixels to mask by giving alpha value of 255, changing global alpha on 
+   *    maskImage will multiply 255 by globalAlpha = [0,1]
+   *  @Param: maskPixels: set of coordinates indexed by red value of pixels to be in mask
+   *  @Param: imgData: Uint8clampedArray of all pixel data in scaled image. 
+   *  @Param: alphaValue: the number the user passes in on a slide bar for transparency of the mask
+   */
+  private drawMask(maskPixels: Set<number>, scaledData: ImageData, alphaValue: number): void {
+    //clear old context
+    this.ctx.clearRect(0,0,this.canvas.nativeElement.width, this.canvas.nativeElement.height);
+
+    //  reset global alpha if image was tinted
+    this.ctx.globalAlpha = 1;
+
+    // access all pixels in original mask, scale them up to image on UI width
+    for (let pixel of maskPixels) {
+      let y = this.getY(this.originalData.width, pixel);
+      let x = this.getX(this.originalData.width, pixel, y);
+      this.maskData.data[this.getPixel(x, y, scaledData.width) + 3] = 255;
+    } 
```

Comment:
Is this actually scaling the mask up?  So if the image is 1x1 and the scaled up canvas is 2x2 we need to fill all 4 pixels with the mask.  But I think this just does 1 pixel in the scaled mask.

Instead of trying to manually scale the mask up or down can we instead have the mask the same size as the original image and then have canvas scale it?  This is how you do it for the original image.

Reply:
I thought the same this as well (about it not scaling the mask) and originally I had a loop going through all the pixels to set the scaledXscaled array based on the pixel of maskPixels. But then when I created masks, the mask hung over part of the image not in the mask. The bottom and the right sides of the new mask were too large. I believe when I get the new pixel with the scaled width it scales the pixel to the image shown. 

Instead of scaling them 1:1 I agree that scaling them all together like I did with the mask makes sense. I will change it.

Reply:
The ctx that the mask is drawn on is already scaled from editor initialization in editor.ts, I believe that's why the mask hung over the right and bottom sides.